pkg/objstore/providers/azure/bucket_client.go

```diff
@@ -9,26 +9,30 @@ import (
 	"github.com/go-kit/log"
 	"github.com/thanos-io/objstore"
 	"github.com/thanos-io/objstore/providers/azure"
-	"gopkg.in/yaml.v3"
 )
 
 func NewBucketClient(cfg Config, name string, logger log.Logger) (objstore.Bucket, error) {
+	return newBucketClient(cfg, name, logger, azure.NewBucketWithConfig)
+}
+
+func newBucketClient(cfg Config, name string, logger log.Logger, factory func(log.Logger, azure.Config, string) (*azure.Bucket, error)) (objstore.Bucket, error) {
 	// Start with default config to make sure that all parameters are set to sensible values, especially
 	// HTTP Config field.
 	bucketConfig := azure.DefaultConfig
 	bucketConfig.StorageAccountName = cfg.StorageAccountName
 	bucketConfig.StorageAccountKey = cfg.StorageAccountKey.String()
+	bucketConfig.StorageConnectionString = cfg.StorageConnectionString.String()
 	bucketConfig.ContainerName = cfg.ContainerName
-	bucketConfig.Endpoint = cfg.Endpoint
 	bucketConfig.MaxRetries = cfg.MaxRetries
 	bucketConfig.UserAssignedID = cfg.UserAssignedID
 
-	// Thanos currently doesn't support passing the config as is, but expects a YAML,
-	// so we're going to serialize it.
-	serialized, err := yaml.Marshal(bucketConfig)
-	if err != nil {
-		return nil, err
+	// do not delay retries
+	bucketConfig.PipelineConfig.RetryDelay = -1
```

Comment:
Perhaps a naive question, why do we want to change the default behavior here? From what I can see Mimir doesn't do this. If we are worried about retries taking too long, there is also MaxRetryDelay that could be set.

Reply:
Without the changes from #2741 I get the hang on startup with those 2 example commands. This goes away when I apply #2741.

Testing this branch, I don't see a difference in behavior between the changes here and latest `main`. Commenting this line out (reenabling retry delay) doesn't seem to make a difference as well for me.

Reply:
This not something I see, I do get a hang on main. I also don't see any big effect on max retries:

```
// main f80352606c8e13a0bcc65805e66899e5e664c275 with #2741, hangs > 1 min

// the max retries change and RetryDelay=-1, fails  after 10s

// the mimir config change, fails after 2-3s

// this pr as of right now, fail after 2-3s
```

I still think all of those changes should be done. MaxRetries=20 is too high, so this expoential back of